### PR TITLE
Add message that SRV check was skipped if .well-known is configured

### LINKED ISSUE #25: UI reports "No SRV records" even when no check was made for SRV records
If `/.well-known/matrix/server` exists and contains a server name *with an explicit port*, then no SRV lookup is performed. Nevertheless the UI still reports "No SRV records", which is somewhat confusing.

See `matrix.org` for example: https://federationtester.matrix.org/#matrix.org

### PR DESCRIPTION
**Pull request by hamburghammer**
The backend does not check for the SRV records if a .well-known
configuration with a port declaration was found. It will return an
empty SRVRecords field which leads to the understanding that nothing
was found even though it was intentionally skipped. The newly introduced
message will inform the user that it was skipped because a .well-known
configuration was found.

Tries to resolve #25 

Refs:
- https://github.com/matrix-org/matrix-federation-tester/issues/104

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -404,6 +404,16 @@ let DNSResult = create({
 
   render: function() {
     let j = this.props.json;
+    let wellknown = this.props.wellknown;
+
+    if (wellknown["m.server"].includes(":")) {
```
Comment (erikjohnston) [inline app.js:409]: Technically this should also check if its an IP address too.

Comment (hamburghammer) [reply app.js:409]: If I understand you correctly you think that it should also check for valid:
- IPv4 address
- IPv6 address
- domain name
The current implementation defensively does not work with IPv6, but should I add for all three cases a regex validation?

**Comment by hamburghammer** [comment]
@erikjohnston The check know supports IPv6 addresses as well as IPv4 and domains.

**Review by richvdh — COMMENTED** [review]
hrm. I appreciate the contribution, but I feel like this is the wrong approach - it infers too much about the backend's behaviour based on heuristics.

I'd much prefer it if we could update the backend to return something in the JSON to indicate that SRV lookup was skipped, so that we can use that here.

How's your Golang? :innocent:

**Comment by hamburghammer** [comment]
I just did a rebase for a cleaner commit history.
It now contains the changes to use the new `SRVSkipped` field.

File: app.js [hunk 2]
```diff
@@ -404,6 +404,15 @@ let DNSResult = create({
 
   render: function() {
     let j = this.props.json;
+
+    if (j.SRVSkipped) {
+      return (
+        <div className="dns">
+          <h2>.well-known information found -&gt; SRV record-check skipped</h2>
```
Comment (richvdh) [inline app.js:411]: This isn't really accurate.

```suggestion
          <h2>server name/.well-known result contains explicit port number: no SRV lookup done</h2>
```

Comment (richvdh) [reply app.js:411]: Also: h2 is a bit big: it looks like an error of some kind. Shouldn't this be in a "DNS results" section, with smaller text saying that we skipped SRV lookup?

(The existing `if (j.SRVRecords == null)` logic is wrong too, but there is no need to make the problem worse)

File: app.js [hunk 3]
```diff
@@ -404,6 +404,15 @@ let DNSResult = create({
 
   render: function() {
     let j = this.props.json;
+
+    if (j.SRVSkipped) {
+      return (
```
Comment (richvdh) [inline app.js:409]: why do we omit the data from `Hosts` in this case? There is still useful information there.

**Review by richvdh — CHANGES_REQUESTED** [review]
thank you! I think we can make this more accurate though.

File: app.js (hunk 3 above)
Comment (hamburghammer) [reply app.js:409]: The Hosts that are shown in this case are the IP of the Host found inside the `.well-known` file. This IP is already listed inside the connection results list, I am currently looking to add a table showing the found Hostname and the resolved IP to give at leased some context to the `Address` list.
I have currently not that much time 🙈 but will try to find a slot this week to update this PR

**Comment by hamburghammer** [comment]
@richvdh 
Sorry for my absents, but I finally found some time to work on this PR. Thanks for not closing it in the mean time.
I had to do some changes to the way the DNS results where displayed and I'm not sure if it's the best possible way to do it, but I am quiet happy with the result.

File: app.js [hunk 4]
```diff
@@ -404,60 +404,76 @@ let DNSResult = create({
...
   render: function() {
     let j = this.props.json;
-    if (j.SRVRecords == null) {
-      return (
-        <div className="dns">
-          <h2>No SRV Records</h2>
-        </div>
-      );
+
+    let dnsResultTitle = <> <h3>DNS records for {j.SRVCName}</h3> </>
```
Comment (richvdh) [inline app.js:408]: note that the `<> ... </>` is redundant here:

```suggestion
    let dnsResultTitle = <h3>DNS records for {j.SRVCName}</h3>;
```

File: app.js [hunk 5]
```diff
@@ -468,25 +484,19 @@ let DNSResult = create({
...
-      </>
+      </>;
     }
 
     return (
       <div className="dns">
-        <h2>DNS records for {j.SRVCName}</h2>
+        <h2>DNS results</h2>
+        {dnsResultTitle}
         <div className="table">
```
Comment (richvdh) [inline app.js:494]: why not include this `<div className="table">` inside `records`? (also: it would probably make sense to include `dnsResultTitle` too.)

File: app.js [hunk 6]
```diff
@@ -404,60 +404,76 @@ let DNSResult = create({
...
-      return (
-        <div className="row" key={id}>
-          <div className="col">{record.Target}</div>
-          <div className="col">{record.Port}</div>
-          <div className="col">{record.Priority}</div>
-          <div className="col">{record.Weight}</div>
-        </div>
-      );
-    });
+    let records;
```
Comment (richvdh) [inline app.js:416]: I think it would make more sense to call this `srv_records_table` or something. `records` is very undescriptive.

File: app.js [hunk 7]
```diff
@@ -404,60 +404,76 @@ let DNSResult = create({
...
-      return (
-        <div className="dns">
-          <h2>No SRV Records</h2>
-        </div>
-      );
+
+    let dnsResultTitle = <> <h3>DNS records for {j.SRVCName}</h3> </>
+    if (j.SRVSkipped) {
+        dnsResultTitle = <>
+          <h3>DNS records</h3>
```
Comment (richvdh) [inline app.js:411]: It looks like this will give us ` <h2>DNS results</h2>` immediately followed by `<h3>DNS records</h3>`.  This doesn't seem to make much sense. Why do we need an extra heading in this case?

File: app.js [hunk 8]
```diff
@@ -404,60 +404,76 @@ let DNSResult = create({
...
           return (
             <div className="row" key={id}>
               <div className="col">{address}</div>
             </div>
           );
         });
+
+        return (
+          <>
+            <h3>{host}</h3>
```
Comment (richvdh) [inline app.js:453]: while you're changing this: I think it would make sense to include the `Error` for each `Host` under the same heading, rather than having a separate section for errors.

**Review by richvdh — CHANGES_REQUESTED** [review]
Thank you for revisiting this! I think it's taking shape, but it's a bit hard to follow at the moment. I've made a few suggestions below.

**Comment by richvdh** [comment]
@hamburghammer it would be great if you could find time to put the finishing touches to this PR soon?

File: app.js (hunk 7 above)
Comment (hamburghammer) [reply app.js]: I overlooked it :see_no_evil: It wasn't intended

File: app.js (hunk 6 above)
Comment (hamburghammer) [reply app.js]: I renamed it to `srvRecordsTable` to match the naming style

File: app.js (hunk 5 above)
Comment (hamburghammer) [reply app.js]: To move this into `records` makes sense to me, but I don't understand why it should be included inside the `dnsResultTitle`?

File: app.js (hunk 8 above)
Comment (hamburghammer) [reply app.js]: This wasn't that easy for me to archive. I did a rather large refactoring to archive the composition of the `Host` table to include the `Error` table. It should also not render an empty table if no values are provided for `Host` or `Error`. To implement it I had to create a rather unsexy if/else check chain...

File: app.js [hunk 9]
```diff
@@ -403,91 +403,117 @@ let DNSResult = create({
...
+      );
+    });
 
     return (
       <div className="dns">
-        <h2>DNS records for {j.SRVCName}</h2>
+        <h2>DNS results</h2>
+        {dnsResultTitle}
+        {srvRecordsTable}
         <div className="table">
```
[the remark this answers is not in the document]
Comment (hamburghammer) [reply app.js]: The styling gets applied anyways, but I changed it to make it explicit.

File: app.js [hunk 10]
```diff
@@ -403,92 +403,102 @@ let DNSResult = create({
...
+          </>
+        );
       }
-    });
-    hostErrors = hostErrors.filter((error) => error != undefined)
+    }();
+
+    const hosts = Object.keys(j.Hosts).map((host) => {
+      const addressTable = function() {
+        const addresses = function() {
```
Comment (richvdh) [inline app.js:445]: I think at least one of these layers of `function()`s is redundant, but it's fine I guess.

File: app.js [hunk 11]
```diff
@@ -403,92 +403,102 @@ let DNSResult = create({
...
+    const skippedSrvResultTitle = <p>server name/.well-known result contains explicit port number: no SRV lookup done</p>;
+    const foundSrvResultTitle = <h3>DNS records for {j.SRVCName}</h3>;
+    const srvResultTitle = j.SRVSkipped ? skippedSrvResultTitle : foundSrvResultTitle;
 
-    let hosts = Object.keys(j.Hosts).map((host) => {
-      if (j.Hosts[host].Addrs != null) {
-        return j.Hosts[host].Addrs.map((address, id) => {
+    const srvRecordsTable = function() {
+      if (j.SRVRecords != null) {
+        const recordRows = j.SRVRecords.map((record, id) => {
```
Comment (richvdh) [inline app.js:414]: I've been doing some testing on this, and generally it now looks great for servers with SRV records, and those where the SRV lookup is skipped.

However, there is a third case we're not correctly handling, which is when the SRV lookup is not skipped, but there are no SRV records - see for example the results for `sw1v.org`:

image

In this case, `SRVSkipped` is false, but `SRVCName` is `""` and `SRVRecords` is `null`.

How about moving the logic for the title into this function:

```suggestion
    const srvRecordsTable = function() {
      if (j.SRVSkipped) {
        return <p>server name/.well-known result contains explicit port number: no SRV lookup done</p>;
      } else if (j.SRVRecords == null) {
        return <p>No SRV records found</p>;
      } else { 
        const recordRows = j.SRVRecords.map((record, id) => {
```

(and add `<h3>DNS records for {j.SRVCName}</h3>` further down)

Then `srvResultTitle` and friends can be removed.

**Review by richvdh — CHANGES_REQUESTED** [review]

File: app.js (hunk 11 above)
Comment (hamburghammer) [reply app.js:414]: This looks defensively weird. Thx for catching it :see_no_evil:

**Review by richvdh — APPROVED** [review]
This looks great now. Thank you so much for your perseverance!
